File: src/com/interviewQuestion_DSA_Java/Divide_Conquer_9/Merge_sort.java
package com.interviewQuestion_DSA_Java.Divide_Conquer_9;

public class Merge_sort {
    public static void printArr(int arr[]) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    // Main function to perform merge sort
    private static void mergeSort(int[] arr, int si, int ei) {
        // Base case: if start index is greater than or equal to end index, return
        if (si >= ei) {
            return;
        }

        // Calculate the middle index
        int mid = si + (ei - si) / 2;

        // Recursively sort the left and right halves
        mergeSort(arr, si, mid);
        mergeSort(arr, mid + 1, ei);

        // Merge the sorted halves
        merge(arr, si, mid, ei);
    }

    // Function to merge two sorted halves
    public static void merge(int arr[], int si, int mid, int ei) {
        // Create a temporary array to store the merged result
        //left(0,3) =4 right(4,6) -> 6+0+1
        int temp[] = new int[ei - si + 1];
        int i = si;  // idx for 1 sorted part
        int j = mid + 1; //idx for 2nd sorted part
        int k = 0;  /// idx for temp;

        while (i <= mid && j <= ei) {
            if (arr[i] < arr[j]) {
                temp[k++] = arr[i++];
            } else {
                temp[k++] = arr[j++];
            }
        }
// Copy the remaining elements of the right half
        //for leftover elements of 1st sorted part
        while (i <= mid) {
            temp[k++] = arr[i++];
        }
//for leftover elements of 2nd sorted part
        while (j <= ei) {
            temp[k++] = arr[j++];
        }
// Copy elements from temp back to arr
        //copy temp to original array
        for (k = 0, i = si; k < temp.length; k++, i++) {
            arr[i] = temp[k];
        }
    }
    public static void main(String[] args) {
//        int arr[] = {6,3,9,5,2,8,-2};
        int arr[] = {38, 27, 43, 3, 9, 82, 10};
        mergeSort(arr,0,arr.length-1);
        printArr(arr);

    }
// time O(nlogn)
   // space O(n)

}

File: src/com/interviewQuestion_DSA_Java/Sorting_5/UsingC/MergeSort.c
#include <stdio.h>

// 🔁 Merge two sorted subarrays into one sorted array
void merge(int arr[], int left, int mid, int right) {
    int n1 = mid - left + 1;    // Size of left subarray
    int n2 = right - mid;       // Size of right subarray

    int L[n1], R[n2];           // Temporary arrays

    // Copy data to temp arrays L[] and R[]
    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];

    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;

    // Merge the temp arrays back into arr[]
    while (i < n1 && j < n2) {
        if (L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    // Copy remaining elements of L[], if any
    while (i < n1)
        arr[k++] = L[i++];

    // Copy remaining elements of R[], if any
    while (j < n2)
        arr[k++] = R[j++];
}

// 🔁 Recursive Merge Sort function
void mergeSort(int arr[], int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;

        // Sort first and second halves
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        // Merge the sorted halves
        merge(arr, left, mid, right);
    }
}

// 🖨 Function to print the array
void printArray(int arr[], int size) {
    printf("Array: ");
    for (int i = 0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

// 🔰 Main function
int main() {
    int arr[] = {7, 11, 9, 2, 17, 4};  // 🎯 Input array
    int size = sizeof(arr) / sizeof(arr[0]);

    printf("Original ");
    printArray(arr, size);

    // 🚀 Apply Merge Sort
    mergeSort(arr, 0, size - 1);

    printf("Sorted ");
    printArray(arr, size);

    return 0;
}
